Regression from #2734: dangling reference in array size

### Dafny version

master

### Code to produce this issue

```dafny
method Bug() {
    var zero := 0;
    var a := new int[zero] [];
}
```


### Command to run and resulting output

```code
> dafny run src/Scratch.dfy -t:java

Dafny program verifier finished with 1 verified, 0 errors
/Users/salkeldr/Documents/GitHub/libraries/src/Scratch-java/_System/__default.java:15: error: cannot find symbol
    java.math.BigInteger[] _nw0 = (java.math.BigInteger[]) dafny.TypeDescriptor.BIG_INTEGER.newArray(dafny.Helpers.toIntChecked((_2_zero), "Java arrays may be no larger than the maximum 32-bit signed int"));
                                                                                                                                 ^
  symbol:   variable _2_zero
  location: class __default
/Users/salkeldr/Documents/GitHub/libraries/src/Scratch-java/_System/__default.java:15: error: illegal start of type
    java.math.BigInteger[] _nw0 = (java.math.BigInteger[]) dafny.TypeDescriptor.BIG_INTEGER.newArray(dafny.Helpers.toIntChecked((_2_zero), "Java arrays may be no larger than the maximum 32-bit signed int"));
                                                                                                                                ^
2 errors
Error while compiling Java files. Process exited with exit code 1
```

Similar errors for other language. C# hits an unrelated bug fixed by [this change]( https://github.com/dafny-lang/dafny/pull/2818/files#diff-316173dfd9593f6c318a2d32c464f9fdd946865969792183f0b130f1b4f1a30e), but if that patch is applied the same error shows up.



### What happened?

Used `git bisect` to identify this was a regression introduced by #2734. Looks like some rewriting introduces a dangling reference to `_2_zero` somehow. 

### What type of operating system are you experiencing the problem on?

Mac

--- a/Source/DafnyCore/AST/Statements/Statements.cs
+++ b/Source/DafnyCore/AST/Statements/Statements.cs
@@ -483,7 +483,7 @@ public class TypeRhs : AssignmentRhs, ICloneable<TypeRhs> {
     if (original.ArrayDimensions != null) {
       if (original.InitDisplay != null) {
         Contract.Assert(original.ArrayDimensions.Count == 1);
-        ArrayDimensions = new List<Expression> { original.ArrayDimensions[0] };
+        ArrayDimensions = new List<Expression> { cloner.CloneExpr(original.ArrayDimensions[0]) };
         InitDisplay = original.InitDisplay.ConvertAll(cloner.CloneExpr);
       } else {
         ArrayDimensions = original.ArrayDimensions.Select(cloner.CloneExpr).ToList();

--- a/Test/git-issues/github-issue-3343.dfy
+++ b/Test/git-issues/github-issue-3343.dfy
@@ -1,0 +1,6 @@
+// RUN: %baredafny run "%s" -t:java > "%t"
+// RUN: %diff "%s.expect" "%t"
+method Bug() {
+    var zero := 0;
+    var a := new int[zero] [];
+}

--- a/Test/git-issues/github-issue-3343.dfy.expect
+++ b/Test/git-issues/github-issue-3343.dfy.expect
@@ -1,0 +1,2 @@
+
+Dafny program verifier finished with 1 verified, 0 errors

